File: frontend/src/lib/innerBlock.ts
export type InnerAlign = 'left' | 'center' | 'right';

export interface InnerColumn {
  widthPct: number | null;
  align: InnerAlign;
  body: string;
}

const WIDTH_RE = /^\[(\d+(?:\.\d+)?)%\]\s*/;

export function parseInnerLine(line: string): InnerColumn | null {
  const trimmed = line.trim();
  if (!trimmed.startsWith('-')) return null;

  let rest = trimmed.slice(1).trim();
  if (!rest) return { widthPct: null, align: 'center', body: '' };

  let widthPct: number | null = null;
  const widthMatch = rest.match(WIDTH_RE);
  if (widthMatch) {
    widthPct = parseFloat(widthMatch[1]);
    rest = rest.slice(widthMatch[0].length);
  }

  let align: InnerAlign = 'center';
  if (rest.startsWith('[[')) {
    align = 'left';
    rest = rest.slice(2).trimStart();
  } else if (rest.startsWith(']]')) {
    align = 'right';
    rest = rest.slice(2).trimStart();
  }

  return { widthPct, align, body: rest };
}

export function parseInnerBlock(source: string): InnerColumn[] {
  const columns: InnerColumn[] = [];
  for (const line of source.split('\n')) {
    const col = parseInnerLine(line);
    if (col) columns.push(col);
  }
  return columns;
}

export function resolveInnerColumnWidths(columns: InnerColumn[]): number[] {
  const fixed = columns.reduce((sum, col) => sum + (col.widthPct ?? 0), 0);
  const autoCount = columns.filter((col) => col.widthPct === null).length;
  const remainder = Math.max(0, 100 - fixed);
  const autoShare = autoCount > 0 ? remainder / autoCount : 0;
  return columns.map((col) => col.widthPct ?? autoShare);
}

const ALIGN_CLASS: Record<InnerAlign, string> = {
  left: 'inner-col--left',
  center: 'inner-col--center',
  right: 'inner-col--right',
};

export function renderInnerBlockHtml(
  source: string,
  renderMarkdown: (body: string) => string,
): string {
  const columns = parseInnerBlock(source);
  if (columns.length === 0) return '';

  const widths = resolveInnerColumnWidths(columns);
  const gridTemplate = widths.map((width) => `${width}fr`).join(' ');
  const colsHtml = columns.map((col) => {
    const alignClass = ALIGN_CLASS[col.align];
    const innerHtml = renderMarkdown(col.body);
    return `<div class="inner-col ${alignClass}">${innerHtml}</div>`;
  }).join('');

  return `<div class="inner-row" style="grid-template-columns: ${gridTemplate};">${colsHtml}</div>\n`;
}

File: backend/inner_block.py
"""Parse and render ```inner fenced blocks for server-side markdown HTML."""
from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Callable, Literal

InnerAlign = Literal["left", "center", "right"]

_WIDTH_RE = re.compile(r"^\[(\d+(?:\.\d+)?)%\]\s*")
_ALIGN_CLASS: dict[InnerAlign, str] = {
    "left": "inner-col--left",
    "center": "inner-col--center",
    "right": "inner-col--right",
}


@dataclass
class InnerColumn:
    width_pct: float | None
    align: InnerAlign
    body: str


def parse_inner_line(line: str) -> InnerColumn | None:
    trimmed = line.strip()
    if not trimmed.startswith("-"):
        return None

    rest = trimmed[1:].strip()
    if not rest:
        return InnerColumn(width_pct=None, align="center", body="")

    width_pct: float | None = None
    width_match = _WIDTH_RE.match(rest)
    if width_match:
        width_pct = float(width_match.group(1))
        rest = rest[width_match.end() :]

    align: InnerAlign = "center"
    if rest.startswith("[["):
        align = "left"
        rest = rest[2:].lstrip()
    elif rest.startswith("]]"):
        align = "right"
        rest = rest[2:].lstrip()

    return InnerColumn(width_pct=width_pct, align=align, body=rest)


def parse_inner_block(source: str) -> list[InnerColumn]:
    columns: list[InnerColumn] = []
    for line in source.splitlines():
        col = parse_inner_line(line)
        if col is not None:
            columns.append(col)
    return columns


def resolve_inner_column_widths(columns: list[InnerColumn]) -> list[float]:
    fixed = sum(col.width_pct or 0 for col in columns)
    auto_count = sum(1 for col in columns if col.width_pct is None)
    remainder = max(0.0, 100.0 - fixed)
    auto_share = remainder / auto_count if auto_count else 0.0
    return [col.width_pct if col.width_pct is not None else auto_share for col in columns]


def render_inner_block_html(
    source: str,
    render_markdown: Callable[[str], str],
) -> str:
    columns = parse_inner_block(source)
    if not columns:
        return ""

    grid_template = " ".join(
        f"{int(width) if width == int(width) else width}fr"
        for width in resolve_inner_column_widths(columns)
    )
    cols_html = "".join(
        f'<div class="inner-col {_ALIGN_CLASS[col.align]}">{render_markdown(col.body)}</div>'
        for col in columns
    )
    return f'<div class="inner-row" style="grid-template-columns: {grid_template};">{cols_html}</div>\n'
